simulation/test_thread_pool.cpp

```diff
@@ -0,0 +1,215 @@
+	}
+
+	// change the number of active threads and wait for the threads
+	// to settle at the new value
+	void set_active_threads(int target)
+	{
+		std::unique_lock<std::mutex> l(m_mutex);
+		assert(target <= m_pool->num_threads());
+		m_target_active_threads = target;
+		while (m_active_threads != m_target_active_threads)
+		{
+			l.unlock();
+			m_cond.notify_all();
+			std::this_thread::yield();
+			l.lock();
+		}
+	}
+
+	// this is to close a race between a thread exiting and a test checking the
+	// thread count
+	void wait_for_thread_exit(int num_threads)
+	{
+		std::unique_lock<std::mutex> l(m_mutex);
+		m_exit_cond.wait_for(l, std::chrono::seconds(30), [&]()
+		{
+			return m_pool->num_threads() == num_threads;
+		});
+	}
+
+	lt::disk_io_thread_pool* m_pool;
+	std::mutex m_mutex;
+	std::condition_variable m_cond;
+	std::condition_variable m_exit_cond;
+
+	// must hold m_mutex to access
+	int m_active_threads;
+	// must hold m_mutex to access
+	int m_target_active_threads;
+};
+
+TORRENT_TEST(disk_io_thread_pool_idle_reaping)
+{
+	sim::default_config cfg;
+	sim::simulation sim{ cfg };
+
+	test_threads threads;
+	sim::asio::io_service ios(sim);
+	lt::disk_io_thread_pool pool(threads, ios);
+	threads.m_pool = &pool;
+	pool.set_max_threads(3);
+	pool.job_queued(3);
+	TEST_EQUAL(pool.num_threads(), 3);
+	// make sure all the threads are up and settled in the active state
+	threads.set_active_threads(3);
+
+	// first just kill one thread
+	threads.set_active_threads(2);
+	lt::deadline_timer idle_delay(ios);
+	// the thread will be killed the second time the reaper runs and we need
+	// to wait one extra minute to make sure the check runs after the reaper
+	idle_delay.expires_from_now(std::chrono::minutes(3));
+	idle_delay.async_wait([&](lt::error_code ec)
+	{
+		// this is a kludge to work around a race between the thread
+		// exiting and checking the number of threads
+		// in production we only check num_threads from the disk I/O threads
+		// so there are no race problems there
+		threads.wait_for_thread_exit(2);
+		TEST_EQUAL(pool.num_threads(), 2);
+		sim.stop();
+	});
+	sim.run();
+	sim.reset();
+
+	// now kill the rest
+	threads.set_active_threads(0);
+	idle_delay.expires_from_now(std::chrono::minutes(3));
+	idle_delay.async_wait([&](lt::error_code ec)
+	{
+		// see comment above about this kludge
+		threads.wait_for_thread_exit(0);
+		TEST_EQUAL(pool.num_threads(), 0);
+	});
+	sim.run();
+}
+
+TORRENT_TEST(disk_io_thread_pool_abort_wait)
+{
+	sim::default_config cfg;
+	sim::simulation sim{ cfg };
+
+	test_threads threads;
+	sim::asio::io_service ios(sim);
+	lt::disk_io_thread_pool pool(threads, ios);
+	threads.m_pool = &pool;
+	pool.set_max_threads(3);
+	pool.job_queued(3);
+	TEST_EQUAL(pool.num_threads(), 3);
+	pool.abort(true);
+	TEST_EQUAL(pool.num_threads(), 0);
+}
+
+#if 0
+// disabled for now because io_service::work doesn't work under the simulator
+// and we need it to stop this test from exiting prematurely
```

Comment:
I imagine this may be hard to fix in the simulator, especially if the work object is supposed to be released in a separate thread. the main thread would try to fast-forward time to infinity, because it doesn't know of any more events in the future.